gtsfm/utils/verification.py

```diff
@@ -86,38 +91,89 @@ def essential_to_fundamental_matrix(
+    can reject correspondences which are actually inliers.
+
+    References: 
+    - "Fathy et al., Fundamental Matrix Estimation: A Study of Error Criteria", https://arxiv.org/abs/1706.07886
+    - "Hartley, R.~I. et al. Multiple View Geometry in Computer Vision.. Cambridge University Press, Pg 288"
+
+    Algorithm:
+    - l2 = i2Fi1 @ x1
+    - l1 = x2 @ i2Fi1 = [a1, b1, c1]
+    - n1 = EuclideanNorm(Normal(l1)) = sqrt(a1^2 + b1^2)
+    - n2 = EuclideanNorm(Normal(l2))
+    - SED^2 = ( l1 @ x1 )^2 * (1/n1^2 + 1/n2^2)
 
     Args:
-        normalized_coords_i1: normalized coordinates in image i1, of shape Nx2.
-        normalized_coords_i2: corr. normalized coordinates in image i2, of shape Nx2.
-        i2Ei1: essential matrix between two images
+        coordinates_i1: coordinates in image i1, of shape Nx2.
+        coordinates_i2: corr. coordinates in image i2, of shape Nx2.
+        i2Fi1: fundamental matrix between two images.
 
     Returns:
-        Symmetric epipolar distances for each row of the input, of shape N.
+        Symmetric epipolar point-line distances for each row of the input, of shape N.
     """
-    if (
-        normalized_coords_i1 is None
-        or normalized_coords_i1.size == 0
-        or normalized_coords_i2 is None
-        or normalized_coords_i2.size == 0
-    ):
+
+    if coordinates_i1 is None or coordinates_i1.size == 0 or coordinates_i2 is None or coordinates_i2.size == 0:
+        return None
+
+    epipolar_lines_i2 = feature_utils.convert_to_epipolar_lines(coordinates_i1, i2Fi1)  # Ex1
+    epipolar_lines_i1 = feature_utils.convert_to_epipolar_lines(coordinates_i2, i2Fi1.T)  # Etx2
+
+    numerator = np.square(feature_utils.point_line_dotproduct(coordinates_i1, epipolar_lines_i1))
+
+    line_sq_norms_i1 = np.sum(np.square(epipolar_lines_i1[:, :2]), axis=1)
+    line_sq_norms_i2 = np.sum(np.square(epipolar_lines_i2[:, :2]), axis=1)
+
+    return numerator * (1 / line_sq_norms_i1 + 1 / line_sq_norms_i2)
+
+
+def compute_epipolar_distances_sampson(
+    coordinates_i1: np.ndarray, coordinates_i2: np.ndarray, i2Fi1: np.ndarray
+) -> Optional[np.ndarray]:
+    """Compute the sampson distance between corresponding coordinates in two images. Sampson distance is the first
+    order approximation of the reprojection error, and well-estimates the gold standard reprojection error at low
+    values of the reprojection error.
+
+    The squared Sampson distance is computed in OpenCV (Ref 3) as well COLMAP (Ref 4). OpenCV uses the squared sampson
+    error to find inliers in RANSAC (Ref 5).
+
+    References: 
+    - "Fathy et al., Fundamental Matrix Estimation: A Study of Error Criteria", https://arxiv.org/abs/1706.07886
+    - "Hartley, R.~I. et al. Multiple View Geometry in Computer Vision.. Cambridge University Press, Pg 288"
+    - https://github.com/opencv/opencv/blob/29fb4f98b10767008d0751dd064023727d9d3e5d/modules/calib3d/src/five-point.cpp#L375 # noqa: E501
```

Comment:
maybe we can put a number next to the refs? since they are referred to by number?